### PR TITLE
Calc mean and std channel-wise

### PR DESCRIPTION
**Pull request by StefanoD**
Addresses issue [#6718](https://github.com/fchollet/keras/issues/6718).

Someone any ideas how to make the code more concise?

### CODE REVIEW

**Comment by StefanoD** [comment]
It took me surprisingly long to simplify this code and I ended almost at the original code.

**Comment by StefanoD** [comment]
For better auditing. Here is an example code:

```python
import numpy as np

b = np.asarray(range(10 * 10 * 3), dtype=np.float32)
b = b.reshape((3, 10, 10))
print("b:\n{}".format(b))
print("b[0]:\n{}".format(b[0]))
print("np.mean(b[0]):{}".format(np.mean(b[0])))
print("np.mean(b[0]):{}".format(np.mean(b[1])))
print("np.mean(b[0]):{}".format(np.mean(b[2])))
print("np.mean(b, axis=(1, 2)):{}".format(np.mean(b, axis=(1, 2))))

c = b.copy()
c[0] -= np.mean(b[0])
c[1] -= np.mean(b[1])
c[2] -= np.mean(b[2])

mean = np.mean(b, axis=(1, 2), keepdims=True)
print("mean:\n{}".format(mean))

print("mean.shape: {}".format(mean.shape))

b -= mean

print("c:\n{}".format(c))
print("b:\n{}".format(b))

print("Are the same: {}".format((b == c).all()))
```

**Comment by fchollet** [comment]
Please, can someone review this and validate the need for it (or inversely)?

**Comment by StefanoD** [comment]
Maybe @taehoonlee ?
Please, see issue [#6718](https://github.com/fchollet/keras/issues/6718) why the current situation is really wrong.

**Comment by StefanoD** [comment]
I'm providing [here](https://github.com/StefanoD/keras_standardize_test) a CIFAR-10 test.
Results for the first 10 epochs:

Original test without sample-wise centering and std normalization:
```
Epoch 1/200
1562/1562 [==============================] - 43s - loss: 1.8458 - acc: 0.3227 - val_loss: 1.5403 - val_acc: 0.4458
Epoch 2/200
1562/1562 [==============================] - 43s - loss: 1.5577 - acc: 0.4294 - val_loss: 1.3388 - val_acc: 0.5210
Epoch 3/200
1562/1562 [==============================] - 43s - loss: 1.4276 - acc: 0.4849 - val_loss: 1.2335 - val_acc: 0.5679
Epoch 4/200
1562/1562 [==============================] - 43s - loss: 1.3431 - acc: 0.5170 - val_loss: 1.1758 - val_acc: 0.5882
Epoch 5/200
1562/1562 [==============================] - 43s - loss: 1.2701 - acc: 0.5461 - val_loss: 1.1309 - val_acc: 0.6039
Epoch 6/200
1562/1562 [==============================] - 43s - loss: 1.2082 - acc: 0.5706 - val_loss: 1.0696 - val_acc: 0.6211
Epoch 7/200
1562/1562 [==============================] - 43s - loss: 1.1624 - acc: 0.5872 - val_loss: 1.0012 - val_acc: 0.6472
Epoch 8/200
1562/1562 [==============================] - 43s - loss: 1.1147 - acc: 0.6063 - val_loss: 0.9848 - val_acc: 0.6549
Epoch 9/200
1562/1562 [==============================] - 43s - loss: 1.0858 - acc: 0.6157 - val_loss: 0.9526 - val_acc: 0.6635
Epoch 10/200
1562/1562 [==============================] - 43s - loss: 1.0523 - acc: 0.6274 - val_loss: 0.9510 - val_acc: 0.6693
```

Sample-wise centering and std normalization, but with the old code:
```
Epoch 1/200
1562/1562 [==============================] - 43s - loss: 1.9395 - acc: 0.2886 - val_loss: 1.7875 - val_acc: 0.3616
Epoch 2/200
1562/1562 [==============================] - 44s - loss: 1.7861 - acc: 0.3529 - val_loss: 1.6832 - val_acc: 0.4014
Epoch 3/200
1562/1562 [==============================] - 44s - loss: 1.7206 - acc: 0.3790 - val_loss: 1.6160 - val_acc: 0.4309
Epoch 4/200
1562/1562 [==============================] - 44s - loss: 1.6752 - acc: 0.3969 - val_loss: 1.5722 - val_acc: 0.4449
Epoch 5/200
1562/1562 [==============================] - 44s - loss: 1.6383 - acc: 0.4095 - val_loss: 1.5551 - val_acc: 0.4483
Epoch 6/200
1562/1562 [==============================] - 44s - loss: 1.6029 - acc: 0.4260 - val_loss: 1.5038 - val_acc: 0.4640
Epoch 7/200
1562/1562 [==============================] - 44s - loss: 1.5698 - acc: 0.4410 - val_loss: 1.4629 - val_acc: 0.4800
Epoch 8/200
1562/1562 [==============================] - 44s - loss: 1.5515 - acc: 0.4490 - val_loss: 1.4481 - val_acc: 0.4862
Epoch 9/200
1562/1562 [==============================] - 44s - loss: 1.5247 - acc: 0.4607 - val_loss: 1.4268 - val_acc: 0.4975
Epoch 10/200
1562/1562 [==============================] - 44s - loss: 1.5046 - acc: 0.4679 - val_loss: 1.3903 - val_acc: 0.5074
```

Sample-wise centering and std normalization, but with the new code:
```
Epoch 1/200
1562/1562 [==============================] - 44s - loss: 1.8956 - acc: 0.3122 - val_loss: 1.5530 - val_acc: 0.4540
Epoch 2/200
1562/1562 [==============================] - 44s - loss: 1.5676 - acc: 0.4433 - val_loss: 1.3243 - val_acc: 0.5347
Epoch 3/200
1562/1562 [==============================] - 43s - loss: 1.4001 - acc: 0.5087 - val_loss: 1.2222 - val_acc: 0.5774
Epoch 4/200
1562/1562 [==============================] - 44s - loss: 1.2915 - acc: 0.5487 - val_loss: 1.1137 - val_acc: 0.6113
Epoch 5/200
1562/1562 [==============================] - 44s - loss: 1.2144 - acc: 0.5767 - val_loss: 1.1176 - val_acc: 0.6160
Epoch 6/200
1562/1562 [==============================] - 44s - loss: 1.1487 - acc: 0.6003 - val_loss: 1.0413 - val_acc: 0.6391
Epoch 7/200
1562/1562 [==============================] - 44s - loss: 1.1064 - acc: 0.6170 - val_loss: 0.9525 - val_acc: 0.6738
Epoch 8/200
1562/1562 [==============================] - 44s - loss: 1.0597 - acc: 0.6314 - val_loss: 0.9260 - val_acc: 0.6778
Epoch 9/200
1562/1562 [==============================] - 44s - loss: 1.0243 - acc: 0.6462 - val_loss: 0.8652 - val_acc: 0.7073
Epoch 10/200
1562/1562 [==============================] - 44s - loss: 0.9979 - acc: 0.6564 - val_loss: 0.8336 - val_acc: 0.7156
```

The new code is beating the other use cases and the hit rate of the second epoch is even better then the 10th epoch of the old centering and std code.
Hope, this convinces you. This result is congruent with my own data set.

**Comment by t-ae** [comment]
I found `featurewise_center` and `featurewise_std_normalization` do similar, but batchwise normalization.
https://github.com/fchollet/keras/blob/17e073d87e0192abb6192dbfb45a97dc8d63a91e/keras/preprocessing/image.py#L659-L671
https://github.com/fchollet/keras/blob/17e073d87e0192abb6192dbfb45a97dc8d63a91e/keras/preprocessing/image.py#L499-L514

**Comment by StefanoD** [comment]
And as I stated in my issue: My approach is more similar to Tensorflow's approach. And further: You can't have a good statistics with three pixels (three channels) or even one pixel (one channel)! This is what the current situation is and that is the reason, why my approach has a **21% points better** hit rate after ten epochs!

**Comment by taehoonlee** [comment]
The current: an average over adjacent channels (a.k.a `ACROSS_CHANNELS` in caffe)
The proposed: an average of each channel (a.k.a `WITHIN_CHANNEL` in caffe)

It is hard to say which one is better normalization for general 4D tensors. However, the issue codes are only for RGB images. I have hardly seen averaging over RGB channels like the current version. I agree with the proposal as @StefanoD stated and Tensorflow works.

**Comment by StefanoD** [comment]
@fchollet Aren't the reviews and the CIFAR-10 tests not convincing enough? It would be cool, if everybody could benefit from this patch and I don't have to maintain a fork.

**Comment by fchollet** [comment]
The main issue here, is that this change would break backwards compatibility and possible break people's trained models (expecting a different normalization scheme). We cannot just merge this change as is.

**Comment by StefanoD** [comment]
@fchollet I could add an _sample_normalization_type_ parameter. This would be an enum where one can specify _ACROSS_CHANNELS_ or _WITHIN_CHANNEL_.
_ACROSS_CHANNELS_ would be default, so everything stays backwards compatible.

**Comment by StefanoD** [comment]
This issue is still unresolved.

**Comment by fchollet** [comment]
We have recently changed the behavior to normalize using statistics of the whole sample, instead of channel-wise (which is what best matches the docstring)..
